### PR TITLE
MGMT-2030 Add QE Registry Cred to Pull Secret

### PR DESCRIPTION
**Pull request by rollandf**
In order to support pulling image from Pre-Prod container Registry, this patch insert the credentials in the Pull Secret.

The Pull Secret is also saved on the ISO image in `/root/.docker/config.json` for Agent and Assisted-Installer images.

### CODE REVIEW

**Comment by rollandf** [comment]
@gamli75

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -148,7 +149,16 @@ const ignitionConfigFormat = `{
           "name": "root"
       },
       "contents": { "source": "data:,{{.AGENT_MOTD}}" }
-	}{{if .RH_ROOT_CA}},
+	},
+	{
+		"overwrite": true,
+		"path": "/root/.docker/config.json",
```
Comment (romfreiman) [inline internal/bminventory/inventory.go:155]: btw, why didnt we do it with auth?

File: internal/cluster/validations/validations.go [hunk 2]
```diff
@@ -76,6 +76,21 @@ func ParsePullSecret(secret string) (map[string]PullSecretCreds, error) {
 	return result, nil
 }
 
+func AddRHRegPullSecret(secret, rhCred string) (string, error) {
```
Comment (romfreiman) [inline internal/cluster/validations/validations.go:79]: @carbonin fyi

File: deploy/assisted-service.yaml [hunk 3]
```diff
@@ -104,6 +104,8 @@ spec:
               value: "info"
             - name: INSTALL_RH_CA
               value: "false"
+            - name: RH_QA_REG_CRED
```
Comment (ronniel1) [inline deploy/assisted-service.yaml:107]: what is this?
From the name I don't understand the purpose

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -213,6 +223,19 @@ func NewBareMetalInventory(
...
+
+	res := pullSecret
+	if b.Config.RhQaRegCred != "" {
+		ps, err := validations.AddRHRegPullSecret(pullSecret, b.Config.RhQaRegCred)
+		if err != nil {
+			log.Infof("Failed to add RH QA Credentials to Pull Secret: %s", err.Error())
+		}
+		res = ps
+	}
+	return res
```
Comment (eranco74) [inline internal/bminventory/inventory.go:236]: Seems that in case of  an error you just log it and return what you got from validations.AddRHRegPullSecret.
If that is the desired behavior?
If so I'd update the code a bit
```suggestion
	if b.Config.RhQaRegCred != "" {
		pullSecret, err := validations.AddRHRegPullSecret(pullSecret, b.Config.RhQaRegCred)
		if err != nil {
			log.Infof("Failed to add RH QA Credentials to Pull Secret: %s", err.Error())
		}
	}
	return pullSecret
```

File: internal/cluster/validations/validations.go (hunk 2 above)
Comment (ronniel1) [reply internal/cluster/validations/validations.go:79]: how is this related to the validations?

File: openshift/template.yaml [hunk 5]
```diff
@@ -79,6 +79,9 @@ parameters:
 - name: INSTALL_RH_CA
   value: "false"
   required: true
+- name: RH_QA_REG_CRED
+  value: ""
+  required: true
```
Comment (ronniel1) [inline openshift/template.yaml:84]: In integration environment we are not using this so why is this required?

File: internal/cluster/validations/validation_test.go [hunk 6]
```diff
@@ -68,6 +70,12 @@ var _ = Describe("Pull secret validation", func() {
 			err := ValidatePullSecret(validSecretFormat, userName, *authHandler)
 			Expect(err).Should(BeNil())
 		})
+		It("Add RH Reg PullSecret ", func() {
```
Comment (eranco74) [inline internal/cluster/validations/validation_test.go:73]: Add a test for empty RHReg

File: internal/cluster/validations/validations.go [hunk 7]
```diff
@@ -76,6 +76,21 @@ func ParsePullSecret(secret string) (map[string]PullSecretCreds, error) {
...
+	var s imagePullSecret
+	err := json.Unmarshal([]byte(secret), &s)
+	if err != nil {
+		return "", fmt.Errorf("invalid pull secret: %v", err)
```
Comment (eranco74) [inline internal/cluster/validations/validations.go:83]: Consider returning the secret you got instead of empty string.

File: internal/cluster/validations/validations.go [hunk 8]
```diff
@@ -76,6 +76,21 @@ func ParsePullSecret(secret string) (map[string]PullSecretCreds, error) {
...
+	}
+	s.Auths["registry.stage.redhat.io"] = make(map[string]interface{})
+	s.Auths["registry.stage.redhat.io"]["auth"] = base64.StdEncoding.EncodeToString([]byte(rhCred))
+	ps, err := json.Marshal(s)
+	if err != nil {
+		return "", err
```
Comment (eranco74) [inline internal/cluster/validations/validations.go:89]: Same here

File: deploy/assisted-service.yaml (hunk 3 above)
Comment (rollandf) [reply deploy/assisted-service.yaml:107]: Red Hat QA Registry Credentials.
Additional credentials needed in order to pull container images, that are not part of the regular pull secret.
In our case `registry.stage.redhat.io`

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (rollandf) [reply internal/bminventory/inventory.go:155]: What do you mean?

File: internal/bminventory/inventory.go (hunk 4 above)
Comment (rollandf) [reply internal/bminventory/inventory.go:236]: I thought that for this part continuing the flow maybe makes more sense.
What do you think? Fail the create/update of PS?

File: internal/cluster/validations/validations.go (hunk 2 above)
Comment (rollandf) [reply internal/cluster/validations/validations.go:79]: inventory.go will be better?

File: internal/cluster/validations/validations.go (hunk 7 above)
Comment (rollandf) [reply internal/cluster/validations/validations.go:83]: OK

File: internal/cluster/validations/validations.go (hunk 8 above)
Comment (rollandf) [reply internal/cluster/validations/validations.go:89]: OK

File: internal/cluster/validations/validation_test.go (hunk 6 above)
Comment (rollandf) [reply internal/cluster/validations/validation_test.go:73]: Sure, will do.

File: openshift/template.yaml (hunk 5 above)
Comment (rollandf) [reply openshift/template.yaml:84]: Additional credentials needed in order to pull container images, that are not part of the regular pull secret.
In our case registry.stage.redhat.io.
Not needed for Brew.
See Jira for additional explainations.

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (eranco74) [reply internal/bminventory/inventory.go:155]: I guess he meant the assisted-service-client auth.
Net relevant for this PR

File: internal/bminventory/inventory.go [hunk 9]
```diff
@@ -148,7 +149,16 @@ const ignitionConfigFormat = `{
           "name": "root"
       },
       "contents": { "source": "data:,{{.AGENT_MOTD}}" }
-	}{{if .RH_ROOT_CA}},
+	},
+	{
+		"overwrite": true,
+		"path": "/root/.docker/config.json",
+		"mode": 644,
```
Comment (carbonin) [inline internal/bminventory/inventory.go:156]: I think this needs to be provided as decimal, ref: https://github.com/coreos/ignition/blob/master/doc/configuration-v3_1.md

Comment (rollandf) [reply internal/bminventory/inventory.go:156]: Done the same as MOTD from above.

File: deploy/assisted-service.yaml (hunk 3 above)
Comment (rollandf) [reply deploy/assisted-service.yaml:107]: Any suggestion for the name? 
What about REGISTRY_STAGE_CRED?

File: internal/bminventory/inventory.go (hunk 4 above)
Comment (rollandf) [reply internal/bminventory/inventory.go]: Failing the flow in case of error.

File: deploy/assisted-service.yaml (hunk 3 above)
Comment (rollandf) [reply deploy/assisted-service.yaml:107]: Do we really need it as a secret? these are well-know internally creds for pulling images from registry.stage.redhat.io which is available only in VPN

**Comment by carbonin** [comment]
Do we really need to append this pull secret to the existing one? Can we just write what was passed directly to the `.docker/config.json`?
Or put a different way, if we needed the existing auths in the pull secret, how were we providing that to `podman` before writing it to the docker config?

**Comment by carbonin** [comment]
I ask because if we eventually utilize the [new ignition config endpoint](https://github.com/openshift/assisted-service/pull/310) for this, merging file contents will be make that much more difficult. It would be much easier for the user to provide the file content directly.

File: deploy/assisted-service.yaml (hunk 3 above)
Comment (carbonin) [reply deploy/assisted-service.yaml:107]: > Do we really need it as a secret?

Maybe not, I just make a habit of bringing it up when it looks like it might belong there.

As for the name, this doesn't necessarily need to be about stage or qa, it's just additional registry creds, so maybe just `REGISTRY_CREDS`?

**Comment by rollandf** [comment]
> Do we really need to append this pull secret to the existing one? Can we just write what was passed directly to the `.docker/config.json`?
> Or put a different way, if we needed the existing auths in the pull secret, how were we providing that to `podman` before writing it to the docker config?

There were no docker config until now in ISO . All images were pulled from public registries

File: deploy/assisted-service.yaml (hunk 3 above)
Comment (rollandf) [reply deploy/assisted-service.yaml]: Update to `REGISTRY_CREDS`

**Comment by ronniel1** [comment]
> I ask because if we eventually utilize the [new ignition config endpoint](https://github.com/openshift/assisted-service/pull/310) for this, merging file contents will be make that much more difficult. It would be much easier for the user to provide the file content directly.

_@carbonin https://issues.redhat.com/browse/MGMT-2137 touches this issue of moving the responsibility to the user. But for now, we need this.

**Comment by carbonin** [comment]
> But for now, we need this.

Yup, I agree.

> There were no docker config until now in ISO . All images were pulled from public registries

So can we just write the user-provided credentials to the config rather than merging with the pull secret?

**Comment by rollandf** [comment]
What do you mean by: 
> the user-provided credentials

**Comment by carbonin** [comment]
The contents of `REGISTRY_CREDS` env var

**Comment by carbonin** [comment]
Okay, met with @rollandf 

I misunderstood what was going on here. This PR needs both the pull secret to pull from Red Hat's registries as well as the additional credentials to pull from the pre-prod registries.

/lgtm

File: internal/bminventory/inventory.go (hunk 9 above)
Comment (rollandf) [reply internal/bminventory/inventory.go]: Fixed

**Review by gamli75 — APPROVED** [review]
